private-templates-service/server/src/db/providers/InMemoryDBProvider.ts

```diff
@@ -0,0 +1,206 @@
+    this._matchUsers(query).then(users => {
+      users.forEach(user => {
+        updateCount += 1;
+        this._updateUser(user, updateQuery);
+      });
+    });
+    return Promise.resolve({ success: true, result: updateCount });
+  }
+  async updateTemplate(query: TemplateQuery, updateQuery: TemplateQuery): Promise<JSONResponse<Number>> {
+    let updateCount: number = 0;
+    this._matchTemplates(query).then(templates => {
+      templates.forEach(template => {
+        updateCount += 1;
+        this._updateTemplate(template, updateQuery);
+      });
+    });
+    return Promise.resolve({ success: true, result: updateCount });
+  }
+
+  async insertUser(doc: IUser): Promise<void> {
+    if (this._clone(doc).org === doc.org) {
+      console.log("fuck");
+    }
+    return this._insert(doc, this.users);
+  }
+  async insertTemplate(doc: ITemplate): Promise<void> {
+    this._setTimestamps(doc);
+    return this._insert(doc, this.templates);
+  }
+
+  async getUser(query: UserQuery): Promise<IUser[]> {
+    return this._matchUsers(query);
+  }
+
+  async getTemplate(query: TemplateQuery): Promise<ITemplate[]> {
+    return this._matchTemplates(query);
+  }
+
+  // Will be fixed in a while to use JSONResponse
+  async removeUser(query: UserQuery): Promise<boolean> {
+    await this._matchUsers(query).then(users => {
+      users.forEach(user => {
+        this.users.delete(user._id!);
+      });
+    });
+    return Promise.resolve(true);
+  }
+  async removeTemplate(query: TemplateQuery): Promise<boolean> {
+    await this._matchTemplates(query).then(templates => {
+      templates.forEach(template => {
+        this.templates.delete(template._id!);
+      });
+    });
+    return Promise.resolve(true);
+  }
+
+  protected async _matchUsers(query: UserQuery): Promise<IUser[]> {
+    let res: IUser[] = new Array();
+    this.users.forEach(user => {
+      if (this._matchUser(query, user)) {
+        res.push(this._clone(user));
+      }
+    });
+    return Promise.resolve(res);
+  }
+
+  protected async _matchTemplates(query: TemplateQuery): Promise<ITemplate[]> {
+    let res: ITemplate[] = new Array();
+    this.templates.forEach(template => {
+      if (this._matchTemplate(query, template)) {
+        res.push(this._clone(template));
+      }
+    });
+    return Promise.resolve(res);
+  }
+
+  protected _clone<T>(obj: T): T {
+    let cloned: T = JSON.parse(JSON.stringify(obj));
+    return cloned;
+  }
+
+  protected async _insert<T extends ITemplate | IUser>(doc: T, collection: Map<String, T>): Promise<void> {
+    let docToInsert: T = this._clone(doc);
+    this._checkID(docToInsert);
+    if (!collection.has(docToInsert._id!)) {
+      collection.set(docToInsert._id!, docToInsert);
+    } else {
+      logger.error("Object with id: %s already exists. Insertion failed", doc._id!);
+    }
+  }
+
+  // Makes sure that id of the object is set
+  protected _checkID<T extends ITemplate | IUser>(doc: T) {
```

Comment:
This should have a better name that describes what is actually happening, maybe 'upsertID'